### PR TITLE
Complex-valued variables

### PR DESCRIPTION
**Pull request by projekter**
This is a first draft for the ideas mentioned in #212

### CODE REVIEW

File: src/complex.jl [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    imag(x::AbstractVector{<:AbstractMonomial})
+    imag(x::AbstractTerm)
+    imag(x::AbstractPolynomial)
+
+Return the imaginary part of `x` by applying `imag` to all coefficients and variables; for this purpose, every complex-valued
+variable is decomposed into its real- and imaginary parts.
+
+See also [`iscomplex`](@ref iscomplex), [`isimagpart`](@ref isimagpart), [`real`](@ref).
+"""
+Base.imag(x::AbstractVariable) = 0
```
Comment (blegat) [inline src/complex.jl:96]: Maybe `MA.Zero()` instead

File: src/complex.jl [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+"""
+    ordvar(x::Union{AbstractVariable, AbstractVector{<:AbstractVariable}})
+
+Given some (complex-valued) variable that was transformed by conjugation, taking its real part, or taking its
+imaginary part, return the original variable as it was defined by the user.
+
+See also [`conj`](@ref conj), [`real`](@ref), [`imag`](@ref).
+"""
+ordvar(x::AbstractVariable) = x
```
Comment (blegat) [inline src/complex.jl:44]: Where does this name come from?

Comment (projekter) [reply src/complex.jl:44]: "Ordinary variable". But `origvar` ("original variable"), `declvar` (declared variable) - or unabbreviated forms - also make sense.

File: src/complex.jl (hunk 1 above)
Comment (projekter) [reply src/complex.jl]: Done

**Comment by projekter** [comment]
I am still interested, and if you also are, I'll go back and have a look at what I changed in the meantime (since I'm actively using those ideas locally for a package of mine) and update this request.

**Comment by blegat** [comment]
Ok, thanks, I'll have a look at the update

**Comment by projekter** [comment]
Now I updated this to my current state. And I also uploaded my current implementation of all this in DynamicPolynomials.
Note that I did not include an implementation of `Base.abs2`, as it is not clear whether this should be done as `conj(z)*z` or as `real(z)^2+imag(z)^2`, which will give entirely different representations.
Well... you could already debate this for `real(z)`. If `z = z_r + im*z_i`, is this then `z_r` or rather `(z+conj(z))/2`? For polynomials/monomials, this will make a huge difference in how many terms there are.

File: src/complex.jl (hunk 2 above)
Comment (blegat) [reply src/complex.jl:45]: Let's use the unabbreviated

Comment (projekter) [reply src/complex.jl]: Done

File: src/show.jl [hunk 3]
```diff
@@ -22,14 +22,31 @@ Base.show(io::IO, p::TypesWithShow) = show(io, MIME"text/plain"(), p)
...
         print_subscript(io, mime, indices)
     end
+    isrealpart(var) && print(io, "ᵣ")
+    isimagpart(var) && print(io, "ᵢ")
+end
+
+function _show(io::IO, mime::MIME"text/print", var::AbstractVariable)
+    if isconj(var)
+        for c in name(var)
+            print(io, c, '\u0305')
```
Comment (blegat) [inline src/show.jl:42]: Could you add comment for what this is doing ?

File: src/show.jl [hunk 4]
```diff
@@ -22,14 +22,31 @@ Base.show(io::IO, p::TypesWithShow) = show(io, MIME"text/plain"(), p)
...
+        end
+    else
+        print(io, name(var))
+    end
+    isrealpart(var) && print(io, "ᵣ")
```
Comment (blegat) [inline src/show.jl:47]: Same here, what is this doing ?

File: src/operators.jl [hunk 5]
```diff
@@ -322,11 +322,13 @@ function _promote_terms(t1::AbstractTerm{LinearAlgebra.UniformScaling{T}}, t2::A
...
-LinearAlgebra.adjoint(v::AbstractVariable) = v
-LinearAlgebra.adjoint(m::AbstractMonomial) = m
-LinearAlgebra.adjoint(t::AbstractTerm) = _term(LinearAlgebra.adjoint(coefficient(t)), monomial(t))
+LinearAlgebra.adjoint(v::AbstractVariable) = conj(v)
+LinearAlgebra.adjoint(m::AbstractMonomial) = conj(m)
+LinearAlgebra.adjoint(t::AbstractTerm) = _term(LinearAlgebra.adjoint(coefficient(t)), LinearAlgebra.adjoint(monomial(t)))
 LinearAlgebra.adjoint(p::AbstractPolynomialLike) = polynomial(map(LinearAlgebra.adjoint, terms(p)))
 LinearAlgebra.adjoint(r::RationalPoly) = adjoint(numerator(r)) / adjoint(denominator(r))
+LinearAlgebra.hermitian_type(::Type{T}) where {T<:AbstractPolynomialLike} = T
+LinearAlgebra.hermitian(v::AbstractPolynomialLike, ::Symbol) = (@assert(!iscomplex(v)); v)
```
Comment (blegat) [inline src/operators.jl:331]: An assert is when you know it should happen independently of the user input (because it can be disabled), if it depends on the user input, throw an error

File: src/complex.jl [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+iscomplex(::AbstractVariable) = false
+Base.isreal(v::AbstractPolynomialLike) = !iscomplex(v)
+"""
+    isrealpart(x::AbstractVariable)
+
+Return whether the given variable is the real part of a complex-valued variable.
+
+See also [`iscomplex`](@ref iscomplex), [`isimagpart`](@ref isimagpart), [`isconj`](@ref isconj).
+"""
+isrealpart(::AbstractVariable) = false
```
Comment (blegat) [inline src/complex.jl:20]: Add an empty line after

File: src/complex.jl [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    end
+    return false
+end
+Base.conj(x::M) where {M<:AbstractMonomial} =
+    convert(M, reduce(*, conj(var)^exp for (var, exp) in zip(variables(x), exponents(x)); init=constantmonomial(x)))
+Base.conj(x::V) where {V<:AbstractVector{<:AbstractMonomial}} = monovec(conj.(x))
+Base.conj(x::T) where {T<:AbstractTerm} = convert(T, conj(coefficient(x)) * conj(monomial(x)))
+Base.conj(x::P) where {P<:AbstractPolynomial} = nterms(x) == 0 ? x : convert(P, sum(conj(t) for t in x))
+
+real_coefficient_type_polynomial_like(::AbstractPolynomialLike{R}) where {R<:Real} = R
```
Comment (blegat) [inline src/complex.jl:115]:
```suggestion
real_coefficient_type(::AbstractPolynomialLike{R}) where {R<:Real} = R
```

Comment (blegat) [reply src/complex.jl:115]: Why not just `real(coefficient_type(...))` ?

File: src/complex.jl [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            # subs throws an error if it doesn't receive at least one substitution
+            full_version = length(substs) > 0 ? subs(x, substs...) : polynomial(x)
+            # Now everything that is imaginary can be recognized by its coefficient
+            return convert(
+                polynomialtype(full_version, real_coefficient_type_polynomial_like(full_version)),
+                mapcoefficients!($fun, full_version)
+            )
+        end
+        Base.$fun(x::AbstractVector{<:AbstractMonomial}) = map(Base.$fun, x)
+        Base.$fun(x::AbstractPolynomial{T}) where {T} = nterms(x) == 0 ?
```
Comment (blegat) [inline src/complex.jl:132]:
```suggestion
        Base.$fun(x::AbstractPolynomial{T}) where {T} = iszero(x) ?
```

File: src/show.jl (hunk 3 above)
Comment (projekter) [reply src/show.jl:42]: Say we create a complex-valued variable `z`. Then, its conjugate is displayed as bar-z (z with overbar).

File: src/show.jl (hunk 4 above)
Comment (projekter) [reply src/show.jl:47]: and the real/imaginary parts get an "r" or "i" subscript, `z_r` or `z_i`

File: src/complex.jl (hunk 7 above)
Comment (projekter) [reply src/complex.jl:115]: True, because I was not aware that `real` also works on the types directly. But this is much cleaner, so I dropped the ugly function.

File: src/complex.jl [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+real_coefficient_type_polynomial_like(::AbstractPolynomialLike{R}) where {R<:Real} = R
+real_coefficient_type_polynomial_like(::AbstractPolynomialLike{Complex{R}}) where {R<:Real} = R
+# Real and imaginary parts are harder to realize. The real part of a monomial can easily be a polynomial.
+for fun in [:real, :imag]
+    eval(quote
+        function Base.$fun(x::Union{AbstractMonomial,AbstractTerm})
+            # We replace every complex variable by its decomposition into real and imaginary part
+            substs = [var => real(var) + 1im * imag(var) for var in variables(x) if iscomplex(var)]
+            # subs throws an error if it doesn't receive at least one substitution
+            full_version = length(substs) > 0 ? subs(x, substs...) : polynomial(x)
```
[the remark this answers is not in the document]
Comment (projekter) [reply src/complex.jl]: Done

File: src/complex.jl (hunk 8 above)
Comment (projekter) [reply src/complex.jl]: Done

[the remark this answers is not in the document]
Comment (projekter) [reply src/complex.jl]: Can you explain what you mean with "directly"? Do you refer to the last function on the `AbstractPolynomial` or the monomial vector?
If you mean the last function, I guess that you refer to using the polynomial constructor that directly allows to specify a vector of terms or even a function that is applied to the coefficients. Yes, this would be a way to do it more efficiently; however, the real/imaginary part is not just applied to the coefficients, but also to the monomials, and it will expand those monomials a lot (unless we want to include a special type of monomial which corresponds to the real or imaginary part of another monomial). So this expansion would have to taken into account and combined anyway, and then I'm no longer sure if this kind of optimization will really help us a lot...

File: src/show.jl (hunk 3 above)
Comment (blegat) [reply src/show.jl:42]: Nice, adding that as comment in the code could help

File: src/complex.jl [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+function iscomplex(p::AbstractPolynomialLike)
+    for v in variables(p)
+        if !iszero(maxdegree(p, v)) && iscomplex(v)
+            return true
+        end
+    end
+    return false
+end
+Base.conj(x::M) where {M<:AbstractMonomial} =
+    convert(M, reduce(*, conj(var)^exp for (var, exp) in zip(variables(x), exponents(x)); init=constantmonomial(x)))
```
Comment (blegat) [inline src/complex.jl:115]:
```suggestion
    convert(M, reduce(*, conj(var)^exp for (var, exp) in powers(x); init=constantmonomial(x)))
```

File: src/complex.jl [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+# deliver simpler types than the inputs if they were deliberately casted, e.g., term to monomial)
+function iscomplex(p::AbstractPolynomialLike)
+    for v in variables(p)
+        if !iszero(maxdegree(p, v)) && iscomplex(v)
+            return true
+        end
+    end
+    return false
+end
+Base.conj(x::M) where {M<:AbstractMonomial} =
```
[the remark this answers is not in the document]
Comment (blegat) [reply src/complex.jl:114]: Same for the three methods below

File: src/complex.jl [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    convert(M, reduce(*, conj(var)^exp for (var, exp) in zip(variables(x), exponents(x)); init=constantmonomial(x)))
+Base.conj(x::V) where {V<:AbstractVector{<:AbstractMonomial}} = monovec(conj.(x))
+Base.conj(x::T) where {T<:AbstractTerm} = convert(T, conj(coefficient(x)) * conj(monomial(x)))
+Base.conj(x::P) where {P<:AbstractPolynomial} = iszero(nterms(x)) ? x : convert(P, sum(conj(t) for t in x))
+
+# Real and imaginary parts are harder to realize. The real part of a monomial can easily be a polynomial.
+for fun in [:real, :imag]
+    eval(quote
+        function Base.$fun(x::Union{AbstractMonomial,AbstractTerm})
```
Comment (blegat) [inline src/complex.jl:123]:
```suggestion
        function Base.$fun(x::APL)
```

File: src/complex.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (blegat) [reply src/complex.jl]: Actually, since you do `mapcoefficients!` after, it could work, I made an inline suggestion

File: src/complex.jl (hunk 12 above)
Comment (projekter) [reply src/complex.jl]: Yes, this would be a way. Perhaps more memory-hungry over all, but faster.
My initial thoughts were (if I remember correctly) that if I do the substitions on a monomial-level first, this might already lead to some cancellations in the expanded monomials, and very likely to dropping a lot of terms (because only the real or imaginary part is retained). So I expand (= large number of terms) the small parts first, leading to simplifications on this level, then add them all up. If the same function is used on the APL level, everything is expanded first, then potentially dropped by the application of the function. Of course, these are all just guesstimates, for example for random complex-valued coefficients, none of these cancellations will have any effect at all.

**Comment by blegat** [comment]
Looks good, could you rebase and add the comment in the code ? Then I think it's good to merge and we'll make it part of https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/issues/250

**Comment by projekter** [comment]
Done - the complex functionality should also appear somewhere in the documentation. Which are the correct places? I assume
`iscomplex` -> Polynomial
`isrealpart`, `isimagpart`, `isconj`, `ordinary_variable` -> Variable
`degree_complex`, `halfdegree` -> Term
`mindegree_complex`, `minhalfdegree`, `maxdegree_complex`, `maxhalfdegree`, `extdegree_complex`, `exthalfdegree` -> Polynomial

And I guess tests would also be a nice thing - but until there is a concrete implementation merged, this does not seam feasible?

**Comment by blegat** [comment]
Yes, the documentation places look good. You can add tests and change `.github/workflow/ci.yml` so that it uses your branch for DynamicPolynomials as I did in https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/pull/231/commits/ecdd33fb09e4c6491654b4d16d650aea64b84b9d for instance

**Comment by projekter** [comment]
Whew, when trying to implement the tests, I needed to go on the current master. I hope I found all the necessary replacements to the new conventions. Locally, all the tests work apart from two broken ones. One is due to `iszero` not being implemented on `MutableArithmetics.Zero` (since I don't know much/anything about MA, I can't comment on this). The other one is due to a difficulty on parsing incomplete substitutions - which hopefully is not the standard case. This could be mitigated (though it's more a DynamicPolynomials issue), but at some conversion cost.

**Comment by projekter** [comment]
@blegat , is there something that still prevents this from being merged? Everything should be adapted for the current version, the test are working by now with the DP implementation (though you can always argue that there could be more tests), the functions are documented...

**Comment by blegat** [comment]
Thanks, no let's try to push this through the finish line.

File: src/operators.jl [hunk 13]
```diff
@@ -446,17 +446,24 @@ function _promote_terms(
...
-LinearAlgebra.adjoint(m::AbstractMonomial) = m
+LinearAlgebra.adjoint(v::AbstractVariable) = conj(v)
+LinearAlgebra.adjoint(m::AbstractMonomial) = conj(m)
 function LinearAlgebra.adjoint(t::AbstractTerm)
-    return _term(LinearAlgebra.adjoint(coefficient(t)), monomial(t))
+    return _term(adjoint(coefficient(t)), adjoint(monomial(t)))
 end
 function LinearAlgebra.adjoint(p::AbstractPolynomialLike)
-    return polynomial(map(LinearAlgebra.adjoint, terms(p)))
+    return polynomial!(adjoint.(terms(p)))
```
Comment (blegat) [inline src/operators.jl:455]: Why the `!` here ? If the coefficient type is `BigInt` then `conj(x::BigInt) = x`  so we won't get a copy. We don't want `adjoint` to return a polynomial for which the coefficient alias to the same coefficeints as the input.

Comment (projekter) [reply src/operators.jl:455]: Ah, yes. Let me explain how this came into being. When I ran into issues with the adjoints, I searched for why it was taken incorrectly/not at all - as I already implemented the adjoint correctly for `AbstractTerm` a couple of commits ago - and found that now there was this new implementation of `adjoint` for `Term`, which did not conjugate the monomial. I then tried to understand why this implementation was there at all, because (apart from the missing conjugation), it is basically identical to the abstract one, apart from missing the `copy_if_mutable`. So I removed the specialized implementation altogether.
Finally, I then found that the same "duplication" also existed for `APL` and `Polynomial`, which looked to do the same thing, only that the one for `Polynomial` looked more concise - so I took it for the abstract one.
With the understanding about `BigInt`s, I see why the implementation should look like it was before, using `polynomial(map(...`. However, then, I don't understand why the implementations for `Term` and `Polynomial` were "safe" - they also aliased the monomials, and in principle, these (being `AbstractMonomials`) could have `BigInt`s as exponents.
(Actually, to exploit the issue you mention, one would need to particularly call the mutating GMP functions - the usual overloaded operators would not mutate, but create copies...)

Comment (blegat) [reply src/operators.jl]: Good point. Whether the monomials are mutable can be checked with `mutability`. For instance, `DynamicPolynomials`'s monomials are mutable in the commutative case (thanks to the fact that their exponents are stored in a vector which is mutable). So if we don't copy the monomial, this is indeed a bug

Comment (blegat) [reply src/operators.jl]: It seems the copy is done:
https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/blob/90ad74e0bd5085842da6337af297bcbdfa64eab6/src/default_term.jl#L89-L94

Comment (projekter) [reply src/operators.jl]: But is this even called? The previous implementation was
```julia
LinearAlgebra.adjoint(t::Term) = Term(adjoint(coefficient(t)), monomial(t))
```
If `t` is a `Term` and I check `@which Term(adjoint(coefficient(t)), monomial(t))`, then this goes to `default_term.jl:20`, i.e., the default struct constructor. And `monomial(t)` gives the field of the term verbatim. The functionality to give the mutable copy if necessary is there, but it was also not invoked in the previous implementation, I'd say.

Given that the implementation for `adjoint` now also invokes `adjoints` on the monomials, which defaults to `conj`, which is supposed to copy if necessary, I would say every implementation of `adjoint` is correct.
On the other hand, I checked that `conj` indeed does copy for all implementations. But there seem to be a few loose ends:
https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/blob/7223c0c7dbde02836de2bb9a30888095fb9af1b5/src/complex.jl#L141-L150
In the real case, I should now replace the `x` by `MA.copy_if_mutable(x)`, if I'm not mistaken?
And in principle, even the default implementation of `conj` for `AbstractVariable` should be changed similarly, because who knows if there'll ever be a mutable variable? Same for the real part.

Finally, the implementation in `DynamicPolynomials.MonomialVector` reads
```julia
Base.conj(x::MonomialVector) = MonomialVector(conj.(x.vars), x.Z) # TODO: do we want to alias Z or copy Z?
```
But given that `mutability` on the `MonomialVector` says `IsNotMutable()`, this is fine and the TODO can be removed?

Comment (blegat) [reply src/operators.jl]: Yes, at the moment, it's still unsafe to assume that operations on multivariate polynomials always return a copy that's completely independent from its input. It's a known issue and you should call `copy_if_mutable` to be safe. I think however that for the main operators `+`, `-`, etc... it should be safe so maybe there are just a few corner cases to fix. Fill free to open issues or PRs to fix these

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -42,6 +42,11 @@ jobs:
             ${{ runner.os }}-test-${{ env.cache-name }}-
             ${{ runner.os }}-test-
             ${{ runner.os }}-
+      - name: DP#complex
```
Comment (blegat) [inline .github/workflows/ci.yml:45]: If you remove this then we'll still be able to see if the tests work in the DynamicPolynomials tests right ? Then we should do this because I can't merge and release with this

Comment (projekter) [reply .github/workflows/ci.yml]: Yes, I think so. In DP, I still go to this branch, but once it's merged, I can also remove this customization there.

File: src/operators.jl (hunk 13 above)
Comment (projekter) [reply src/operators.jl]: Ok, then for the time being, I just added the `copy_if_mutable` where I mentioned it.

File: src/complex.jl [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+end
+function Base.conj(x::V) where {V<:AbstractVector{<:AbstractMonomial}}
+    return isreal(x) ? MA.copy_if_mutable(x) : monomial_vector(conj.(x))
+end
+function Base.conj(x::T) where {T<:AbstractTerm}
+    return isreal(x) ? MA.copy_if_mutable(x) :
+           convert(T, conj(coefficient(x)) * conj(monomial(x)))
+end
+function Base.conj(x::P) where {P<:AbstractPolynomial}
+    return iszero(nterms(x)) || isreal(x) ? MA.copy_if_mutable(x) :
```
Comment (blegat) [inline src/complex.jl:149]:
```suggestion
    return iszero(x) || isreal(x) ? MA.copy_if_mutable(x) :
```

File: src/complex.jl [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+           convert(P, sum(conj(t) for t in x))
```
Comment (blegat) [inline src/complex.jl:150]:
```suggestion
           convert(P, polynomial([conj(t) for t in x]))
```

File: src/complex.jl [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+    exthalfdegree
+
+"""
+    iscomplex(x::AbstractVariable)
+
+Return whether a given variable was declared as a complex-valued variable (also their conjugates are complex, but their real
+and imaginary parts are not).
+By default, all variables are real-valued.
+"""
+iscomplex(::AbstractVariable) = false
```
Comment (blegat) [inline src/complex.jl:22]: Would it be possible to remove this function and use `!isreal` instead ? `iscomplex` is a bit ambiguous since the set of real numbers is a subset of the set of complex numbers. So in a sense, `iscomplex` should be true for real ones as well.

File: src/complex.jl [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+# Real and imaginary parts are harder to realize. The real part of a monomial can easily be a polynomial.
+for fun in [:real, :imag]
+    eval(
+        quote
+            function Base.$fun(x::_APL)
+                # Note: x may also be a polynomial; we could handle this case separately by performing the replacement in each
+                # individual term, then adding them all up. This could potentially lower the overall memory requirement (in case
+                # the expansions of the individual terms simplify) at the expense of not being able to exploit optimizations that
+                # subst can do by knowing the full polynomial.
+                iszero(nterms(x)) &&
```
Comment (blegat) [inline src/complex.jl:162]:
```suggestion
                iszero(x) &&
```

File: src/complex.jl [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+                    return zero(polynomial_type(x, real(coefficient_type(x))))
+                # We replace every complex variable by its decomposition into real and imaginary part
+                substs = [
+                    var => real(var) + 1im * imag(var) for
+                    var in variables(x) if iscomplex(var)
+                ]
+                # subs throws an error if it doesn't receive at least one substitution
+                full_version =
+                    isempty(substs) ? polynomial(x) : subs(x, substs...)
```
Comment (blegat) [inline src/complex.jl:171]: Splatting `Vector`s always scares me since that will compile a different method for any number of substitutions. I use a vector version of the substitution instead.

File: src/complex.jl (hunk 17 above)
Comment (projekter) [reply src/complex.jl:22]: While I fully understand your argument, isn't the issue the same with `isreal`? One could also assume that `isreal(x) = iszero(imag(x))`, but this is not the way it is implemented. Maybe this is even an argument for renaming (and therefore not using the Base function) them into `is_declared_real` and/or `is_declared_complex` or something similar?

File: src/complex.jl (hunk 19 above)
Comment (projekter) [reply src/complex.jl]: Done

File: src/complex.jl (hunk 17 above)
Comment (projekter) [reply src/complex.jl:22]: I thought some more about your comment. My previous response was more focused on pointing out the "for all values" character of these methods even more, but when applied to a raw variable, it should be quite clear that the behavior is indeed "for all values" - so adding unnecessary "declared" function name complexity does not make much sense.

My choice to have a function `iscomplex` that enables the complex-valued behavior came from two ideas:
- The default behavior is reals-only. So an additional function has to be provided that "enables" the behavior. This is different in mindset (but nothing else) to having a function that provides more of a legacy not-supported switch.
- I needed to do some special handling for the complex case. So actually, I rarely (actually never) use `isreal`, but `iscomplex` arises much more often. This is more of a convenience argument - but strictly speaking, `!isreal` is still shorter than `iscomplex`.

So if you wish, I'll remove `iscomplex` completely and make everything default to `!isreal`.

**Review by blegat — APPROVED** [review]
Thanks for all the work, it's been a requested feature for a long time and the implementation looks neat. We can merge once ci passes
